platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ExportMigrationContextImpl.java

```diff
@@ -0,0 +1,278 @@
+            new TreeMap<>();
+
+    private final ExportMigrationReportImpl report;
+
+    private final ZipOutputStream zos;
+
+    private volatile OutputStream currentOutputStream;
+
+    /**
+     * Creates a new migration context for an export operation.
+     *
+     * @param report     the migration report where to record warnings and errors
+     * @param migratable the migratable this context is for
+     * @param zos        the output stream for the zip file being generated
+     * @throws IllegalArgumentException if <code>report</code>, <code>migratable</code>, <code>zos</code>
+     *                                  is <code>null</code>
+     * @throws java.io.IOError          if unable to determine ${ddf.home}
+     */
+    ExportMigrationContextImpl(MigrationReport report, Migratable migratable, ZipOutputStream zos) {
+        super(report,
+                migratable,
+                ExportMigrationContextImpl.validateNotNull(migratable, "invalid null migratable")
+                        .getVersion());
+        Validate.notNull(zos, "invalid null zip output stream");
+        this.report = new ExportMigrationReportImpl(report, migratable);
+        this.zos = zos;
+    }
+
+    private static <T> T validateNotNull(T t, String msg) {
+        Validate.notNull(t, msg);
+        return t;
+    }
+
+    @Override
+    public ExportMigrationReportImpl getReport() {
+        return report;
+    }
+
+    @Override
+    public Optional<ExportMigrationEntry> getSystemPropertyReferencedEntry(String name,
+            BiPredicate<MigrationReport, String> validator) {
+        Validate.notNull(name, "invalid null system property name");
+        Validate.notNull(validator, "invalid null validator");
+        final ExportMigrationSystemPropertyReferencedEntryImpl me = systemProperties.get(name);
+
+        if (me != null) {
+            return Optional.of(me);
+        }
+        final String val = System.getProperty(name);
+
+        if (!validator.test(report, val)) {
+            return Optional.empty();
+        } else if (val == null) {
+            report.record(new MigrationException(Messages.EXPORT_SYSTEM_PROPERTY_NOT_DEFINED_ERROR,
+                    name));
+            return Optional.empty();
+        } else if (val.isEmpty()) {
+            report.record(new MigrationException(Messages.EXPORT_SYSTEM_PROPERTY_IS_EMPTY_ERROR,
+                    name));
+            return Optional.empty();
+        }
+        final ExportMigrationSystemPropertyReferencedEntryImpl sprop =
+                new ExportMigrationSystemPropertyReferencedEntryImpl(this, name, val);
+
+        systemProperties.put(name, sprop);
+        return Optional.of(sprop);
+    }
+
+    @Override
+    public ExportMigrationEntry getEntry(Path path) {
+        final ExportMigrationEntryImpl e = new ExportMigrationEntryImpl(this, path);
+
+        return entries.computeIfAbsent(e.getPath(), p -> e);
```

Comment:
❓ With the current logic, won't we end up creating a new `ExportMigrationEntryImpl` even if it does exist in the map?

Reply:
Yeah! but that shouldn't be a big deal. I just do not want to pull the logic for getting the path and cleaning up outside of the entry. This is much easier at the cost of one simple object being thrown away if and only if it happens that getEntry() is called a second time.